File: python/dijkstra.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-

import priority_queue

def dijkstra(graph, start):
    '''
    @property
    graph[] = (dst, weight)
    start: int
    '''
    V = len(graph)
    dist = [float('inf') for i in range(V)]
    used = [False for i in range(V)]
    q = priority_queue()
    q.push((0, start), 0)
    while not q.empty():
        _, _, (d, current_node) = q.pop()
        if used[current_node]: continue
        used[current_node] = True
        dist[current_node] = d
        for next_node, edge_cost in graph[current_node]:
            if dist[next_node] <= d + edge_cost:
                continue
            q.push((d+edge_cost, next_node), -(d+edge_cost))
    return dist

def dijkstra_with_path(graph, s, t):
    '''
    @property
    graph[] = (dst, weight)
    s: int
    '''
    V = len(graph)
    dist = [float('inf') for i in range(V)]
    prev = [-1 for i in range(V)]
    q = priority_queue()
    q.push((s, 0, s), 0)
    while not q.empty():
        _, _, (prev_node, d, current_node) = q.pop()
        if prev[current_node] != -1:
            continue
        prev[current_node] = prev_node
        dist[current_node] = d
        for next_node, edge_cost in graph[current_node]:
            if dist[next_node] <= d + edge_cost:
                continue
            q.push((current_node, d+edge_cost, next_node), -(d+edge_cost))

    # build path
    path = []
    p = t
    while prev[p] != p:
        path.append(p)
        p = prev[p]
    path.append(p)
    return dist[t], list(reversed(path))

def main():
    print("dijkstra!")

if __name__ == "__main__":
    main()

File: c++/dijkstra.cpp
#include <vector>
#include <utility>
#include <tuple>
#include <queue>

struct edge {
    edge(int _to, int _cost) : to(_to), cost(_cost) {};
    int to, cost;
};
typedef std::vector<std::vector<edge>> Graph;
std::vector<int> dijkstra(const Graph &graph, const int s){
    typedef std::pair<int, int> P;
    const int INF = 1e9;
    int V = graph.size();
    std::vector<int> dist(V, INF);
    std::vector<bool> used(V, false);
    std::priority_queue<P, std::vector<P>, std::greater<P>> q;
    q.push(P(0, s));
    while(!q.empty()){
        P current = q.top(); q.pop();
        int d = current.first;
        int current_node = current.second;
        if(used[current_node]) continue;
        used[current_node] = true;
        dist[current_node] = d;
        for(edge e: graph[current_node]){
            if(dist[e.to] <= d + e.cost) continue;
            q.push(P(d+e.cost, e.to));
        }
    }
    return dist;
}

std::pair<int, std::vector<int>> dijkstra(const Graph& graph, const int s, const int t){
    typedef std::tuple<int, int, int> P;
    const int INF = 1e9;
    int V = graph.size();
    std::vector<int> dist(V, INF);
    std::vector<int> prev(V, -1);
    std::priority_queue<P, std::vector<P>, std::greater<P>> q;
    q.push(std::make_tuple(s, 0, s));
    while(!q.empty()){
        P current = q.top(); q.pop();
        int prev_node    = std::get<0>(current);
        int d            = std::get<1>(current);
        int current_node = std::get<2>(current);
        if(dist[current_node] != INF) continue;
        prev[current_node] = prev_node;
        dist[current_node] = d;
        for(edge e: graph[current_node]){
            if(dist[e.to] <= d + e.cost) continue;
            q.push(std::make_tuple(current_node, d+e.cost, e.to));
        }
    }

    // build path
    std::vector<int> path = {t};
    for(int j=t; prev[j] != j; j = prev[j]){
         path.push_back(j);
    }
    return std::make_pair(dist[t], reverse(path.begin(), path.end()));
}
